pkg/kv/kvserver/rangefeed_restarter.go

```diff
@@ -0,0 +1,227 @@
+			},
+			func(r *Replica, p rangefeed.Processor) {
+				r.disconnectRangefeedWithErr(p,
+					kvpb.NewError(kvpb.NewRangeFeedRetryError(kvpb.RangeFeedRetryError_REASON_REPLICA_REMOVED)))
+			})
+	})
+}
+
+// startCatchup bumps number of active catch-ups counter and returns a function
+// thant must be called when catchup finished to release quota.
+func (p *RangefeedRestarter) startCatchup() (done func()) {
+	atomic.AddInt64(&p.activeCatchups, 1)
+	once := sync.Once{}
+	return func() {
+		once.Do(func() {
+			atomic.AddInt64(&p.activeCatchups, -1)
+		})
+	}
+}
+
+// run performs a gradual restart of processors that don't match desired type.
+func run(
+	ctx context.Context,
+	s *Store,
+	scheduler bool,
+	waitRestartQuota func(ctx context.Context) error,
+	restartFn func(*Replica, rangefeed.Processor),
+) {
+	log.VInfof(ctx, 2, "attempting to update processors to new type of %t", scheduler)
+	var restartedProcessors int
+	defer func() {
+		log.VInfof(ctx, 2, "restarted processors with different type: %d", restartedProcessors)
+	}()
+
+	ids := findProcessorsOfType(s, !scheduler)
+	log.VInfof(ctx, 2, "found rangefeed processors with non matching config to restart %d", len(ids))
+	for _, id := range ids {
+		r := s.GetReplicaIfExists(id)
+		if r == nil {
+			continue
+		}
+		// Skip removed processors without waiting.
+		if r.getRangefeedProcessor() == nil {
+			continue
+		}
+		if err := waitRestartQuota(ctx); err != nil {
+			// This is only the case of context cancelled and store shutting down or
+			// when settings were changed to disable gradual restart completely.
+			log.VInfof(ctx, 2, "stopping gradual processor restart: %s", err)
+			return
+		}
+		// Get and recheck processor after waiting because it might have been
+		// removed or replaced.
+		proc := r.getRangefeedProcessor()
+		if proc == nil {
+			continue
+		}
+		_, scheduled := proc.(*rangefeed.ScheduledProcessor)
+		if scheduled != scheduler {
+			restartFn(r, proc)
+			restartedProcessors++
+		}
+	}
+}
+
+// findProcessorsOfType finds range ids of replicas in the store that have
+// rangefeed processor that match desired scheduler configuration filter.
+func findProcessorsOfType(s *Store, scheduler bool) []roachpb.RangeID {
+	var ids []roachpb.RangeID
+	s.rangefeedReplicas.Lock()
+	defer s.rangefeedReplicas.Unlock()
+	for rangeID, schedulerID := range s.rangefeedReplicas.m {
+		if !scheduler && schedulerID == 0 || scheduler && schedulerID > 0 {
+			ids = append(ids, rangeID)
+		}
+	}
+	return ids
+}
+
+// waitRestartQuota waits until enough catchup quota is available but not less
+// than pacing interval.
+// cfgChangedC is monitored for configuration changes
+// necessary catchup quota threshold is determined as the difference between
+// activeCatchups and concurrentCatchupLimit * minQuotaPercent to ensure
+// restarted rangefeeds not hitting store catchup limiter.
+func waitRestartQuota(
```

Comment:
I get that we did this for testability, but it's a bit unfortunate that we're doing this separately from the main loop, because we could otherwise reuse most of the stuff here between iterations. That would also allow runtime between waits to count towards the pacing interval. One alternative would be to use a testing knob instead for the test needs.

It's unclear if this really matters though, in terms of overhead, I'll leave it to your judgement.

Reply:
Folded that into main method. Test has to do some more prep.